test/unit/runner/plugins/TestPluginServiceTest.php

```diff
@@ -62,6 +62,14 @@ class TestPluginServiceTest extends TestCase
             ]
         ];
 
+    /**
+     * @before
+     */
+    public function init(): void
+    {
+        define('CONFIG_PATH', '/');
```

Comment:
@gitromba can we create the bootstrap file on generis instead and just adapt the phpunit.xml of the projects instead? 

https://github.com/oat-sa/extension-tao-test/pull/350/files#diff-9546e6d2d4589ae0a3714a03ca312e1bR1

Otherwise we will have lot of code duplication,

wdyt?

Reply:
Good point, @gabrielfs7! 

I like explicitly defined dependencies especially if they all are in the same place because they are making me confident that my tests are well isolated and I should expect what I see here and no more. This brings although some duplication as you mentioned.

A second concern against the bootstrap and global config is that they are global (surprise :) which makes the tests more coupled to a to each other and their specific values. You may expect a certain value for one constant and perhaps you require something else in another test. Finding such bug could also be painful especially if you are lacking of knowledge on downstream application components (classes) internals (see previous paragraph for reasoning).

When we talk about a bigger picture of TAO extensions, majority of them don't own `phpunit.xml` configurations. One would naturally assume engineers already used to this state and so their awareness needs to be raised to a new approach.

Furthermore the CI pipeline has to be updated since phpunit configuration happens by on-the-fly generated phpunit.xml configuration. This is not a big deal though but yet has to be carried out. And it has to be done before merging the _bootstrap_ solution to avoid broken pipelines.

Reply:
Hello @gitromba , 
This class seems to be constructed using static method to "inject" dependencies, and this is the root causes. (Unfortunately) we cannot refactor old legacy software to have proper testing. 
It means that your platform has to be installed, and you are testing as a group, e.g. an integration test. Even if the test is a unit one. 

Imho, we should move this test to integration to avoid duplication. (I know we have different point of view with @gabrielfs7 😄 )

I agree with solution of @gabrielfs7 but it is out of scope of your PR ;)

Hope it helps

Reply:
Hey @siwane and thanks for your input! I'm glad to observe something from multiple POVs :)

Unit tests belong to extension and as long as we have the same approach applied across all extensions I'm ok with that. This is required to maintain the CI workflow efficiently.

Are you suggesting to fall back to [this solution](https://github.com/oat-sa/extension-tao-test/pull/350/files#diff-9546e6d2d4589ae0a3714a03ca312e1bR1)?

Reply:
Effectively, something like that, load sample instead of real platform must be the good solution.
However,  by using this sample you are pointing to an empty config folder that may not work with your usage. 
A more elaborate loading has to be investigated because [this part](https://github.com/oat-sa/generis/blob/101d7edbaec77bbe35a23bc9610a31b9dbabcb91/common/oatbox/AbstractRegistry.php#L43) requires to have "installed" config. 
In other term, the constant definition will be solved but it is really used by the code you are testing.

Reply:
@gitromba 

Thanks for being proactive and fixing this. This is being a pain for the CI :) 

I understand your point and agree, but it will cause us to add those constants and other dependencies every time we want to unit test something that is legacy and has some static calls inside. Refactor code in TAO during development is not always feasible. We have code spread on libraries, other extensions (some times customer extensions), static calls, etc.

@camille I think those are not integration, but (social) unit tests and if we move as integration we will lose their original intent and tests will become inflated / mixed.

The extensions should be independent from the others, but this is not the reality and these cases will occur more often. The tests today are really dependent on generis TestCase class already.

Also, If you let those constants in the tests and the dependency on those are removed, how the developers will know they can remove the constant from the tests? Maybe it will remain there for a long time and developers will be afraid to remove it.

**A suggestion**

We could have an agreement to avoid static calls / use those global constants directly in the classes (and refactor if possible). Until there, I am not sure we should force the developer to fix this legacy limitation on every test. Steps:

- Create an agreement with developers to avoid use GLOBAL constants directly.
  - Refactor the code to remove those when possible (during development).
- Create a "minimal stability bootstrap" to allow tests and avoid code replication.
- Do evaluation how to remove those for good and create refactoring tasks.
- Remove those.
- Remove the bootstrap.

Alternatively we can continue doing this on every occurrence, but I am not sure it will be fixed one day. 

I can say to you we are avoiding use those constants / static calls whenever is possible.

Reply:
Just for sharing. Unit test without mocking all dependencies can be considered `Sociable Unit test` and it is preferred by Martin Fowler: https://martinfowler.com/bliki/UnitTest.html

Reply:
Interesting thoughts from Mr Fowler. I have to find my balls first then I would challenge his POV on mocking 3rd party services: `If talking to the resource is stable and fast enough for you then there's no reason not to do it in your unit tests.` How one can guarantee stability of a 3rd party service? Just a rhetorical question.

Reply:
> @gitromba
> 
> Thanks for being proactive and fixing this. This is being a pain for the CI :)
> 
> I understand your point and agree, but it will cause us to add those constants and other dependencies every time we want to unit test something that is legacy and has some static calls inside. Refactor code in TAO during development is not always feasible. We have code spread on libraries, other extensions (some times customer extensions), static calls, etc.
> 
> @Camille I think those are not integration, but (social) unit tests and if we move as integration we will lose their original intent and tests will become inflated / mixed.
> 
> The extensions should be independent from the others, but this is not the reality and these cases will occur more often. The tests today are really dependent on generis TestCase class already.
> 
> Also, If you let those constants in the tests and the dependency on those are removed, how the developers will know they can remove the constant from the tests? Maybe it will remain there for a long time and developers will be afraid to remove it.
> 
> **A suggestion**
> 
> We could have an agreement to avoid static calls / use those global constants directly in the classes (and refactor if possible). Until there, I am not sure we should force the developer to fix this legacy limitation on every test. Steps:
> 
> * Create an agreement with developers to avoid use GLOBAL constants directly.
>   
>   * Refactor the code to remove those when possible (during development).
> * Create a "minimal stability bootstrap" to allow tests and avoid code replication.
> * Do evaluation how to remove those for good and create refactoring tasks.
> * Remove those.
> * Remove the bootstrap.
> 
> Alternatively we can continue doing this on every occurrence, but I am not sure it will be fixed one day.
> 
> I can say to you we are avoiding use those constants / static calls whenever is possible.

+1 for avoiding direct constant dependencies. What would be an efficient solution instead?

Reply:
> Effectively, something like that, load sample instead of real platform must be the good solution.
> However, by using this sample you are pointing to an empty config folder that may not work with your usage.
> A more elaborate loading has to be investigated because [this part](https://github.com/oat-sa/generis/blob/101d7edbaec77bbe35a23bc9610a31b9dbabcb91/common/oatbox/AbstractRegistry.php#L43) requires to have "installed" config.
> In other term, the constant definition will be solved but it is really used by the code you are testing.

Having a global config makes tests harder to maintain. Devs must make sure the configuration satisfies their own use case and respects all other. The latter one might be tricky to validate with global constants.

If we commit ourselves to _bootstrap_ way I recommend proceeding step by step as @gabrielfs7 advised and assess the progress continuously so we could fine-tune our approach gradually and remain pragmatical.

Reply:
@gabrielfs7 @siwane I've reopened [the previous PR](https://github.com/oat-sa/extension-tao-test/pull/350) with the bootstrap solution. Let's continue there.

Reply:
I understand is harder to maintain, but we do not have choice. If we are defining constant by constant, we may causes consistency issue. 
This is because we are relying on global constant, the perfect solution should be to remove it and avoid to use static method to load dependencies from constructor.
For example, if u are defining a constant `CONFIG_PATH`, it must point to expected value. In ur test case, it is not a problem because you will just load an empty registry. But by defining it to `/` you are loading your root system folder as config folder, and I do not recommend that.
As final word, if the bootstrap includes value that fit your case, it may not be good for another testing. 
Once again, the problem of dependency injection is the root cause, we must be able to load parameter instead of constants.
We should ask Technical Enabler to, at least, have a common vision.
Wdyt ?

Reply:
As a Technical Enabler I've made the changes :)
- tao-test: https://github.com/oat-sa/extension-tao-test/pull/350/files
- generis: https://github.com/oat-sa/generis/pull/832